### ISSUE TITLE
Not differentiable with the given tensors error when trying to train multi-input neural network

### ISSUE THREAD

**Issue by AditMeh** [raskr/rust-autograd#39]
@rusbridger and I are trying to train a multi-input siamese neural network with this structure, however, I get the error that this isn't differentiable.

My inputs x1, x2 and y are of shape [batch, 105, 105, 3], [batch, 105, 105, 3] and [batch, 1] respectively. The type is all f32.

I am running this on Ubuntu WSL.

The error I end up getting is:

`thread 'main' panicked at 'Not differentiable with given tensor(s).`

These are my layers, I import them into the training function below this block of code

Here is the full repo:
https://github.com/UTMIST/oneshot-rs

```rust

/// Compute convolutional layer with max-pooling.
pub fn conv_pool<'g>(x: Tensor<'g>, w: Tensor<'g>, b: Tensor<'g>) -> Tensor<'g> {
    let g = x.graph();
    let y1 = g.conv2d(x, w, 0, 1) + b;
    let y2 = g.relu(y1);
    x.graph().max_pool2d(y2, 2, 0, 2)
}

/// Compute the final convolutional layer.
pub fn conv_final<'g>(x: Tensor<'g>, w: Tensor<'g>, b: Tensor<'g>) -> Tensor<'g> {
    let g = x.graph();
    let y1 = g.conv2d(x, w, 0, 1) + b;
    x.graph().relu(y1)
}

// Load inputs.
pub fn inputs(g: &Graph<f32>) -> (Tensor, Tensor, Tensor) {
    let x1 = g.placeholder(&[-1, 1, 105, 105]);
    let x2 = g.placeholder(&[-1, 1, 105, 105]);
    let y = g.placeholder(&[-1, 1]);
    (x1, x2, y)
}

/// Compute the final sigmoid layer.
pub fn sigmoid_layer<'g>(x: Tensor<'g>, params: &[Tensor<'g>]) -> (Tensor<'g>, Tensor<'g>) {
    let g = x.graph();
    let z1 = conv_pool(x, params[0], params[6]);
    let z2 = conv_pool(z1, params[1], params[7]);
    let z3 = conv_pool(z2, params[2], params[8]);
    let twin_2_z4 = conv_final(z3, params[3], params[9]);
    let flattened = g.reshape(twin_2_z4, &[-1, 256 * 6 * 6]); // flatten
    let dense = flattened.graph().matmul(params[4], flattened) + params[10];
    (dense, g.sigmoid(flattened))
}

```

This is my training function, I just modified the CNN mnist example a bit by adding some of my own layers. This block of code imports the functions from the above code block.

```rust
pub fn train(train_x1: NDArr, train_x2: NDArr, train_y: NDArr) {

    ag::with(|g| {
        let rng = array::ArrayRng::<f32>::default();
        macro_rules! rand_normal {
            ($arr: expr) => {
                g.variable(rng.random_normal(&$arr, 0., 0.1));
            };
        }
        macro_rules! zeroes {
            ($arr: expr) => {
                g.variable(array::zeros(&$arr));
            };
        }

        // Weights/Biases for convolutional layers.
        let w1 = rand_normal!([64, 1, 10, 10]);
        let w2 = rand_normal!([128, 1, 7, 7]);
        let w3 = rand_normal!([128, 1, 4, 4]);
        let w4 = rand_normal!([256, 1, 4, 4]);
        let b1 = zeroes!([1, 64, 105, 105]);
        let b2 = zeroes!([1, 128, 42, 42]);
        let b3 = zeroes!([1, 128, 18, 18]);
        let b4 = zeroes!([1, 256, 6, 6]);

        // Weights/Biases for dense layers.
        let w5 = rand_normal!([4096, 256 * 6 * 6]);
        let w6 = rand_normal!([4096, 1]);
        let b5 = zeroes!([4096, 1]);

        // Collect parameters and add to adam_state.
        let params = &[w1, w2, w3, w4, w5, w6, b1, b2, b3, b4, b5];
        let adam_state = adam::AdamState::new(
            params
                .iter()
                .map(|v| v.get_variable_array().unwrap())
                .collect::<Vec<_>>()
                .as_slice(),
        );

        // Load inputs and compute sigmoid layers.
        let (x1, x2, y) = layers::inputs(g);
        let (_dense_1, sig_1) = layers::sigmoid_layer(x1, params);
        let (_dense_2, sig_2) = layers::sigmoid_layer(x2, params);

        // Siamese Distance
        let pre_weighted_l1_dist = g.abs(sig_2 - sig_1);
        let weighted_l1_dist = g.matmul(w6, pre_weighted_l1_dist);
        let final_prediction = weighted_l1_dist;
        let loss = g.sigmoid_cross_entropy(final_prediction, y);
        let grads = &g.grad(&[&loss], params);
        let _update_ops: &[Tensor] =
            &adam::Adam::default().compute_updates(params, grads, &adam_state, g);

        // println!("{:?}", );
        for epoch in 0..1 {
            let x1_batch = train_x1.slice(s![0..500, .., .., ..]).into_dyn();
            let x2_batch = train_x2.slice(s![0..500, .., .., ..]).into_dyn();
            let y_batch = train_y.slice(s![0..500, ..]).into_dyn();

            println!("{:?}", x1_batch.shape());
            println!("{:?}", x2_batch.shape());
            println!("{:?}", y_batch.shape());

            g.eval(
                _update_ops,
                &[x1.given(x1_batch), x2.given(x2_batch), y.given(y_batch)],
            );
            println!("finish epoch {}", epoch);
        }
    })
}
```

**Comment by raskr** [comment]
Thank you for the report.
`params[5]` does not seem to be used to compute the loss. So the panic occurs I think.
